### ISSUE TITLE
Add urllib exception handlers to download_cached file

### ISSUE THREAD

**Issue by Mixih** [coala/coala#3332]
A failure to download a cached file because the server was sleeping or something shouldn't make coala throw a unhanded bear exception. These should be caught and processed as log warnings.

**Comment by jayvdb** [comment]
Please no.
You had a requests rewrite .  It was good.  Writing our own http(s) code using the low level urllib is not easy or wise.  If was was, requests would be a small wrapper.  Rather, requests was so big that urllib3 was created.

**Comment by Mixih** [comment]
should we rewrite in requests then @jayvdb ?

**Comment by jayvdb** [comment]
Yes.

**Comment by Mixih** [comment]
Alright, personal task list:
1. Rewrite in requests(urllib is hell)
2. Write exception handlers
3. Fix bugs

**Comment by Mixih** [comment]
Shallwe set a sane timeout too so it doesn't hang for hours if you're not connect to the internet?

**Comment by jayvdb** [comment]
Yes, and note that requests has two timeouts, one for connecting and the other for in-transit delays.  On CI, I usually do 30 for both.  The sites we are using are quite large scale systems, so lower values should be ok.

**Comment by jayvdb** [comment]
Small update.  We had a pretty serious regression in the exception handling: https://github.com/coala/coala/issues/3803  , resulting in https://github.com/coala/coala-bears/issues/1461

I think we should add some retry logic to the fetch process : https://github.com/coala/coala/issues/3804

> A failure to download a cached file because the server was sleeping or something shouldn't make coala throw an unhanded bear exception. These should be caught and processed as log warnings.

The fetch is being used in the bears `setup_dependencies`.  Was an exception in there resulting in an unhandled exception. That sounds bad.

If the file cant be downloaded, the bear could be disabled.  I think coala has similar logical if the bear has some other type of problem.

IMO it should be an error if `setup_dependencies` failed.

And then, https://github.com/coala/coala/issues/3087 should give us the correct exit code.

**Comment by jayvdb** [comment]
The requests rewrite is done, and https://github.com/coala/coala/issues/3804 should add the retry.

This issue should remain open to ensure that a failed HTTP fetch disables the bear instead with a warning, and doesnt cause `coala-ci` to fail while fetching.
It should generate a warning, and possibly return a non-zero exit status at the end of the coala-ci job.
